### ISSUE TITLE
IP-based whitelisting when utilizing Istio ingress gateway

### ISSUE THREAD

**Issue by Mirko-Weiler** [gardener/gardener#5630]
**How to categorize this issue?**

/area control-plane networking security\
/kind enhancement\
/priority normal

<br />

**What would you like to be added**:

### High-level summary

We propose the addition of IP-based access control at the shoot ingress layer, utilizing the **already built-in**  authorization policy framework by Istio ingress gateways when relying on `ManagedIstio` [feature gate](https://github.com/gardener/gardener/blob/77e09dd522d7bfb682cc35d0be437201b606061b/docs/usage/istio.md) in gardenlet. 

### Detailed

Shoot API server access can be publicly exposed through Istio ingress gateways. With [this](https://github.com/gardener/gardener/blob/77e09dd522d7bfb682cc35d0be437201b606061b/docs/proposals/08-shoot-apiserver-via-sni.md) feature an ingress gateway has been added to facade access to Shoot API servers by SNI-based routing. Ingress gateway represents an envoy proxy by Istio, providing exposition ([`Gateway`](https://istio.io/latest/docs/reference/config/networking/gateway/)) and routing ([`VirtualServce`](https://istio.io/latest/docs/reference/config/networking/virtual-service/)) of traffic at the shoot layer. `Gateway` and `VirtualService` are created in the `<shoot-namespace>`.

In order to introduce IP-based access control we propose to extend the shoot resource by incorporating portions of Istio's [`AuthorizationPolicy`](https://istio.io/latest/docs/reference/config/security/authorization-policy/). The below illustrates an authorization policy restricting access through IP whitelisting:

```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: <name>
  namespace: <shoot-namespace>
spec:
  selector:
    matchLabels:
      <label>
  action: ALLOW
  rules:
  - from:
    - source:
        ipBlocks: [<IP/CIDR>]
        remoteIpBlocks: [<IP/CIDR>]
```
Please note proxies between client and gateway may dillute origin IP and require to set `X-Forwarded-For` or proxy protocol; this requires usage of `remoteIpBlocks`. Utilizing `externalTrafficPolicy: Local` may persist origin IP, allowing usage of `ipBlocks`. `AuthorizationPolicy` live at the `<shoot-namespace>` and hence are only applicable to workloads in this namespace. 

Hence, the proposed shoot resource may be structured as follow:
```yaml
# fields ignored for brevity
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: <name>
  namespace: <garden-namespace>
spec:
  kubernetes:
     kubeAPIServer:
      accessControl:
        action: ALLOW | DENY
        rules:
        - from:
            - source:
                ipBlocks: [<IP/CIDR>]
                remoteIpBlocks: [<IP/CIDR>]
```

### References
Similar issue has been raised [here](https://github.com/gardener/gardener/issues/3505).

<br />

**Why is this needed**:
<br />
Tightening security- and regulatory frameworks require clients to harden security mechanisms of k8s cluster and containers. Just recently the BSI (BSI - Bundesamt für Sicherheit in der Informationstechnik) has first-time issued APP4.4 Kubernetes of the [IT-Grundschutz-Bausteine](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Grundschutz/IT-GS-Kompendium_Einzel_PDFs_2022/Zip_Datei_Edition_2022.html;jsessionid=3B6E0C991D3945F155CEE475387A4D93.internet462) Kompendium lifting access control mechanisms for state/configuration changes to K8s clusters to "must-comply" provisions for public entities. Likewise, NIST (National Institute of Standards and Technology, USA) recommends enforcement of "least priviledge" principle ([AC-6](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r5.pdf#page=63&zoom=100,124,588)) or access enforcement ([AC-3](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r5.pdf#page=50&zoom=100,117,389)), aiming to restrict access to system resources to authorized clients only.

In summary, IP-based access control at API-Server layer provide for authorized state changes of K8s clusters and hence adress the tightening security policies and surging demand for access control mechanisms by customers.

We are more than happy to provide a PR for this enhancement.

CC @majst01 @Gerrit91 @mwennrich

**Comment by timebertt** [comment]
/remove-lifecycle rotten

@SimonKienzler @breuerfelix can you comment here on the state of the ACL extension?

**Comment by breuerfelix** [comment]
We developed a Gardener Extension based on EnvoyFilters for IP allowlisting.  

https://github.com/stackitcloud/gardener-extension-acl

All design decisions and restrictions we faced are described in the README :). 
TL;DR; only allowlisting works due to how the VPN traffic is handled in Gardener.  

We would love to have some feedback on the extension! Either in here, Slack on as a GitHub Issue.

**Comment by timebertt** [comment]
/close
in favor of https://github.com/stackitcloud/gardener-extension-acl
